### ISSUE TITLE
Improve resilience when client can not connect?

### ISSUE THREAD

**Issue by bitbrain** [BetterCloud/vault-java-driver#39]
Right now _(v. 1.2.0)_ there will be thrown a `VaultException` when calling `vault.logical().write(...)` when the client can not connect. The problem is that, for instance, a token (passed by the VaultConfig) may not be valid anymore (because it is expired or configuration on the server has changed), or the Vault server might be offline for a few seconds which leads to the same outcome (depending on the timeouts).

My question is: should this driver take the responsibility (e.g. by using Hystrix) to ensure reconnecting properly/recovering? Or should I write a "wrapper" around this driver to ensure connectivity and retrying? I don't know if this would be out-of-scope for this project.

:speech_balloon: Discussion appreciated.

**Comment by jentfoo** [comment]
As a user of the library my vote would be for:

> write a "wrapper" around this driver to ensure connectivity and retrying

For us, this is not an issue, and I want failures at startup to be loud because they most likely point to a configuration flaw rather than a service issue.  I would prefer this library is light weight, small, and does only what is necessary to talk to vault.

If complex retry logic, or circuit tripping logic (like hystrix) is desired, then it seems trivial to put as an additional layer within the application.

**Comment by steve-perkins** [comment]
Ultimately, I would like for the core library to be useful for as many people as possible in as many use cases as possible.  To that end, it should stay light and simple... but also avoid rigid decisions that would make it incompatible with common frameworks.

We implemented basic retry logic, following in the pattern already established by HashiCorp with the Ruby driver.  However, I would rather have circuit breaking and more advanced techniques like that remain external to the library.  If our current design makes that more difficult for any reason, then that's fair game to look at.

**Comment by MyRealityCoding** [comment]
> However, I would rather have circuit breaking and more advanced techniques like that remain external to the library. If our current design makes that more difficult for any reason, then that's fair game to look at. 

I started already implementing a wrapper around this library (by using also Hystrix) but I stumbled upon some problems. One particular problem is [this line in the Logical implementation](https://github.com/BetterCloud/vault-java-driver/blob/master/src/main/java/com/bettercloud/vault/api/Logical.java#L67). Ideally I would like to react on HTTP response codes, for example:
* **401 Unauthorized** to know that I need to authenticate again if a token gets invalid while my application is running)
* **404 Not Found** the current path does not exist on Vault -> fall back on a different path or do any other action.

Right now `vault.logical().write(...)` will just throw a `VaultExeption`. I could potentially read out the exception message, grep out the HTTP response and do proper handling afterwards, my question is if there are plans to improve on how response information is presented by this library when doing requests. (it would be even nice if `VaultException` contains the status code itself)
